From 6bf67137ef1ee5cd70c842b014c322b7deaf994b Mon Sep 17 00:00:00 2001
From: Subodh Iyengar <subodh@fb.com>
Date: Thu, 8 Aug 2019 11:38:29 -0700
Subject: [PATCH] Reject zero length handshake records.

Summary:
Zero length (all padding) handshake are forbidden by RFC. Allowing
these was a regression in D13754697 (2c6f78a).

This is a partial fix for CVE-2019-11924

Reviewed By: xybu

Differential Revision: D16285100

fbshipit-source-id: 05a19d31ad74601ce89156a0e59517aaad8dd928
---
 fizz/record/EncryptedRecordLayer.cpp     | 2 +-
 fizz/record/test/EncryptedRecordTest.cpp | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/fizz/record/EncryptedRecordLayer.cpp b/fizz/record/EncryptedRecordLayer.cpp
index 7fbc3d30..a56b7625 100644
--- a/fizz/record/EncryptedRecordLayer.cpp
+++ b/fizz/record/EncryptedRecordLayer.cpp
@@ -147,7 +147,7 @@ folly::Optional<TLSMessage> EncryptedReadRecordLayer::read(
           static_cast<ContentTypeType>(msg.type)));
   }
 
-  if (!msg.fragment) {
+  if (!msg.fragment || msg.fragment->empty()) {
     if (msg.type == ContentType::application_data) {
       msg.fragment = folly::IOBuf::create(0);
     } else {
diff --git a/fizz/record/test/EncryptedRecordTest.cpp b/fizz/record/test/EncryptedRecordTest.cpp
index d3d10511..15307cd2 100644
--- a/fizz/record/test/EncryptedRecordTest.cpp
+++ b/fizz/record/test/EncryptedRecordTest.cpp
@@ -180,7 +180,7 @@ TEST_F(EncryptedRecordTest, TestAllPaddingHandshake) {
         expectSame(buf, "0123456789");
         return getBuf("16000000");
       }));
-  EXPECT_NO_THROW(read_.read(queue_));
+  EXPECT_ANY_THROW(read_.read(queue_));
 }
 
 TEST_F(EncryptedRecordTest, TestNoContentType) {
